File: 378/segment_tree.cpp
#include "segment_tree.h"

#include <cmath>
#include <iostream>

SegmentTree::SegmentTree(size_t size) : mSize(size), mTree(4*size)
{
}

SegmentTree::SegmentTree(const std::vector<ll>& values, size_t size) : mSize(size), mTree(4*size)
{
    _build(values, mRoot, 0, size - 1);
}

void SegmentTree::_build(const std::vector<ll>& values, size_t v, size_t ts, size_t te)
{
    if (ts == te)
        mTree[v] = values[ts];
    else
    {
        size_t tm = (ts + te) / 2;
        _build(values, left(v), ts, tm);
        _build(values, right(v), tm + 1, te);
        mTree[v] = mTree[left(v)] + mTree[right(v)];
    }
}

ll SegmentTree::get(size_t index)
{
    return _range_sum(index, index, mRoot, 0, mSize - 1);
}

void SegmentTree::update(size_t index, ll value, bool isIncrease)
{
    _update(index, value, mRoot, 0, mSize - 1, isIncrease);
}

void SegmentTree::_update(size_t index,
        ll value,
        size_t v,
        size_t ts,
        size_t te,
        bool isIncrease)
{
    if (ts == te)
    {
        mTree[v] = value + (isIncrease ? mTree[v] : 0);
    }
    else
    {
        size_t tm = (ts + te) / 2;
        if (index <= tm)
        {
            _update(index, value, left(v), ts, tm, isIncrease);
        }
        else
        {
            _update(index, value, right(v), tm + 1, te, isIncrease);
        }
        mTree[v] = mTree[left(v)] + mTree[right(v)];
    }
}

ll SegmentTree::range_sum(size_t start, size_t end)
{
    return _range_sum(start, end, mRoot, 0, mSize - 1);
}

ll SegmentTree::_range_sum(size_t start, size_t end, size_t v, size_t ts, size_t te)
{
    if (start > end)
        return 0;
    if (start == ts && end == te)
        return mTree[v];
    size_t tm = (ts + te) / 2;
    return _range_sum(start, std::min(end, tm), left(v), ts, tm) +
            _range_sum(std::max(start, tm + 1), end, right(v), tm + 1, te);
}

File: 378/segment_tree.py
def left(node):
    return 2 * node


def right(node):
    return 2 * node + 1


class SegmentTree:
    def __init__(self, max_size, values=None):
        self.root = 1
        self.ts = 0
        self.te = max_size - 1
        self.max_size = max_size
        self.tree = [0] * (4 * max_size)
        if values:
            self._build(values, self.root, self.ts, self.te)

    def _build(self, values, v, ts, te):
        if ts == te:
            self.tree[v] = values[ts]
        else:
            tm = (ts + te) // 2
            self._build(values, left(v), ts, tm)
            self._build(values, right(v), tm + 1, te)
            self.tree[v] = self.tree[left(v)] + self.tree[right(v)]

    def increase(self, index, by):
        self._update(index, by, self.root, self.ts, self.te, is_increase=True)

    def update(self, index, value):
        self._update(index, value, self.root, self.ts, self.te)

    def _update(self, index, value, v, ts, te, is_increase=False):
        if ts == te:
            self.tree[v] = value + self.tree[v] if is_increase else 0
        else:
            tm = (ts + te) // 2
            if tm >= index: 
                self._update(index, value, left(v), ts, tm)
            else:
                self._update(index, value, right(v), tm + 1, te)
            self.tree[v] = self.tree[left(v)] + self.tree[right(v)]

    def get(self, index):
        return self.range_sum(index, index)

    def range_sum(self, start, end):
        return self._range_sum(start, end, self.root, self.ts, self.te)

    def _range_sum(self, start, end, v, ts, te):
        if start > end:
            return 0
        if ts == start and te == end:
            return self.tree[v]
        else:
            tm = (ts + te) // 2
            return self._range_sum(start, min(tm, end), left(v), ts, tm)\
                    + self._range_sum(max(start, tm + 1), end, right(v), tm + 1, te)